journal/src/main/java/io/zeebe/journal/file/MappedJournalSegmentWriter.java

```diff
@@ -105,72 +133,97 @@ public void append(final JournalRecord record) {
               nextIndex, record.index()));
     }
 
-    final int recordStartPosition = buffer.position();
-    lastEntry = write(buffer, record);
-    index.index(lastEntry, recordStartPosition);
-  }
+    final MutableDirectBuffer writeBuffer = new UnsafeBuffer();
+    final int startPosition = buffer.position();
+    final int metadataLength = serializer.getMetadataLength();
 
-  /**
-   * Create and writes a new JournalRecord with the given index,asqn and data to the buffer. After
-   * the method returns, the position of buffer will be advanced to a position were the next record
-   * will be written.
-   */
-  private JournalRecord write(
-      final ByteBuffer buffer, final long index, final long asqn, final DirectBuffer data) {
+    final JournalIndexedRecord indexedRecord =
+        new JournalIndexedRecordImpl(record.index(), record.asqn(), record.data());
+    checkCanWrite(buffer, indexedRecord);
```

Comment:
It would be nice in the future if for expected exceptions we don't throw but rather handle these error cases properly. I get that's out of scope, but something we should take care of in the future.